### PR TITLE
Add ddp_cpu backend for testing ddp without GPUs

### PR DESCRIPTION
**Pull request by neggert**
New feature: DDP on CPU. This can be used for distributed training with out GPUs, or to test/debug DDP on single machine without GPUs.. Since PyTorch already makes good use of multiple cores under the hood, this *will not* provide any speedup over normal CPU training if you're only using a single node.

API changes:

* New distributed backend: `distributed_backend="ddp_cpu"`
* New `Trainer` argument: `num_processes`. Controls the number or processes per node. Is currently only used by `ddp_cpu`.

Still need to update the documentation, but I wanted to get some eyes on this before I got too far. Right now everything if functional as far as I can tell, and I've added a new test that covers this feature.

### CODE REVIEW

**Comment by tullie** [comment]
Have you considered just making ddp work with cpu, when GPUS=0 or GPUS=None? I think if I was a user that would be the intuitive thing for ddp to do without GPUs.

File: pytorch_lightning/core/lightning.py [hunk 1]
```diff
@@ -855,11 +855,15 @@ def init_ddp_connection(self):
         try:
             root_node = os.environ['SLURM_NODELIST'].split(' ')[0]
         except Exception:
-            root_node = '127.0.0.2'
+            root_node = '127.0.0.1'
```
Comment (williamFalcon) [inline pytorch_lightning/core/lightning.py]: why the IP change?

File: pytorch_lightning/core/lightning.py [hunk 2]
```diff
@@ -855,11 +855,15 @@ def init_ddp_connection(self):
...
 
         root_node = self.trainer.resolve_root_node_address(root_node)
         os.environ['MASTER_ADDR'] = root_node
-        torch_distrib.init_process_group('nccl', rank=proc_rank, world_size=world_size)
+        if self.trainer.on_gpu:
```
Comment (williamFalcon) [inline pytorch_lightning/core/lightning.py]: why the different backends?

File: pytorch_lightning/overrides/data_parallel.py [hunk 3]
```diff
@@ -99,7 +99,14 @@ def forward(self, *inputs, **kwargs):  # pragma: no cover
                 output = self.gather(outputs, self.output_device)
         else:
             # normal
-            output = self.module(*inputs, **kwargs)
+            # output = self.module(*inputs, **kwargs)
+            # lightning (ddp_cpu)
+            if self.module.training:
```
Comment (williamFalcon) [inline pytorch_lightning/overrides/data_parallel.py]: why this change?

File: pytorch_lightning/core/lightning.py (hunk 1 above)
Comment (neggert) [reply pytorch_lightning/core/lightning.py]: I wish I knew. 127.0.0.1 works, 127.0.0.2 doesn't.

File: pytorch_lightning/core/lightning.py (hunk 2 above)
Comment (neggert) [reply pytorch_lightning/core/lightning.py]: NCCL only works for GPU. The PyTorch docs recommend using GLOO for CPU training: https://pytorch.org/docs/stable/distributed.html#backends

File: pytorch_lightning/overrides/data_parallel.py (hunk 3 above)
Comment (neggert) [reply pytorch_lightning/overrides/data_parallel.py]: This is copying the changes you made above for the case where `self.device_ids` is None. I think previously, we would never hit this code.

**Comment by neggert** [comment]
@tullie As discussed in Slack, I'll change things so that `distributed_backend="ddp"`, `gpus=None` also gives DDP-on-CPU behavior. (`gpus=0` runs on GPU 0 I believe. Don't think we should change that.)

**Comment by Borda** [comment]
@neggert how is it going here? need any help? :rabbit:

**Comment by williamFalcon** [comment]
@neggert can we get this merged so it can go into 0.7.2?

File: pytorch_lightning/trainer/distrib_data_parallel.py [hunk 4]
```diff
@@ -174,44 +174,57 @@ def init_tpu(self):
...
+            elif self.num_gpus > 1:
                 warnings.warn('You requested multiple GPUs but did not specify a backend, e.g.'
                               ' Trainer(distributed_backend=dp) (or ddp, ddp2).'
                               ' Setting distributed_backend=dp for you.')
                 self.use_dp = True
-                self.use_ddp = False
-                self.use_ddp2 = False
+        elif distributed_backend == "dp":
+            if self.num_gpus == 0:
+                pass
```
Comment (Borda) [inline pytorch_lightning/trainer/distrib_data_parallel.py:196]: why these ` if ...: pass` ?

File: pytorch_lightning/trainer/distrib_data_parallel.py [hunk 5]
```diff
@@ -174,44 +174,57 @@ def init_tpu(self):
...
+        elif distributed_backend == "ddp2":
+            if self.num_gpus == 0:
+                pass
+            elif self.num_gpus >= 1:
+                self.use_ddp2 = True
+        elif distributed_backend == "ddp_cpu":
+            if self.num_gpus > 0:
+                m = 'You requested one or more GPUs, but set the backend ' \
+                    'to ddp_cpu. Training will not use GPUs.'
+                warnings.warn(m)
```
Comment (Borda) [inline pytorch_lightning/trainer/distrib_data_parallel.py:221]:
```suggestion
                warnings.warn('You requested one or more GPUs, but set the backend to `ddp_cpu`.'
                              ' Training will not use GPUs.')
```

File: tests/trainer/test_trainer.py [hunk 6]
```diff
@@ -661,3 +661,196 @@ def on_batch_start(self, trainer, pl_module):
     assert not trainer.interrupted
     trainer.fit(model)
     assert trainer.interrupted
+
+
+def test_trainer_config_cpu():
```
Comment (Borda) [inline tests/trainer/test_trainer.py:666]: can we parametrize these test per no-GPU, 1-GPU, multi-GPU?
happy to write a suggestion about it... (ping me back)

File: pytorch_lightning/trainer/distrib_data_parallel.py [hunk 7]
```diff
@@ -171,45 +171,52 @@ def init_tpu(self):
...
-
-            elif distributed_backend is None:
+    def set_distributed_mode(self, distributed_backend):
+        self.use_dp = False
+        self.use_ddp = False
+        self.use_ddp2 = False
+        self.single_gpu = False
+
+        if distributed_backend is None:
+            if self.num_gpus == 0:
```
[the remark this answers is not in the document]
Comment (neggert) [reply pytorch_lightning/trainer/distrib_data_parallel.py:184]: Done

File: pytorch_lightning/trainer/distrib_data_parallel.py [hunk 8]
```diff
@@ -171,45 +171,52 @@ def init_tpu(self):
...
+                self.use_ddp = True
+            elif self.num_gpus > 1:
+                self.use_ddp = True
+                self.num_processes = self.num_gpus
+        elif distributed_backend == "ddp2":
+            if self.num_gpus == 0:
+                return
+            elif self.num_gpus >= 1:
+                self.use_ddp2 = True
+        elif distributed_backend == "ddp_cpu":
```
[the remark this answers is not in the document]
Comment (neggert) [reply pytorch_lightning/trainer/distrib_data_parallel.py:217]: Done

File: pytorch_lightning/trainer/distrib_data_parallel.py (hunk 4 above)
Comment (neggert) [reply pytorch_lightning/trainer/distrib_data_parallel.py:196]: Just for clarity. I wanted to make clear to the future reader that we're considering this case and intentionally doing nothing.

File: tests/trainer/test_trainer.py (hunk 6 above)
Comment (neggert) [reply tests/trainer/test_trainer.py]: Done

File: pytorch_lightning/trainer/distrib_data_parallel.py (hunk 4 above)
Comment (Borda) [reply pytorch_lightning/trainer/distrib_data_parallel.py:196]: I would skip it lol

File: tests/models/test_cpu.py [hunk 9]
```diff
@@ -40,6 +42,31 @@ def test_early_stopping_cpu_model(tmpdir):
     model.unfreeze()
 
 
+@pytest.mark.skipif(platform.system() == "Windows",
+                    reason="Distributed training is not supported on Windows")
+@pytest.mark.skipif((platform.system() == "Darwin" and
+                     version_parse(torch.__version__) < version_parse("1.3.0")),
```
Comment (Borda) [inline tests/models/test_cpu.py:48]: this is not very robust assuming pytorch may go about 10, rise `'1.3.0' > '1.10.0'`
check https://stackoverflow.com/a/11887885/4521646

Comment (neggert) [reply tests/models/test_cpu.py:48]: That's why I'm using the version parsing tool from the `packaging` library. It should handle things like this.

https://packaging.pypa.io/en/latest/version/

File: pytorch_lightning/trainer/distrib_data_parallel.py (hunk 4 above)
Comment (justusschock) [reply pytorch_lightning/trainer/distrib_data_parallel.py:196]: Maybe a comment could achieve the same

**Comment by williamFalcon** [comment]
Is this ready for 0.7.2 or push to 0.7.3?

**Comment by Borda** [comment]
I would leave it for next 0.7.3 @neggert ?

**Comment by neggert** [comment]
Yeah, let's not rush it into a release. 0.7.3 is fine.

File: pytorch_lightning/trainer/distrib_data_parallel.py (hunk 4 above)
Comment (neggert) [reply pytorch_lightning/trainer/distrib_data_parallel.py]: Replaced it with a comment

**Comment by Borda** [comment]
changelog need to be rebased on new release #1419

**Comment by neggert** [comment]
I can reproduce the test failure. I don't think it has anything to do with the specific Python version. It happens randomly about 1 time in 10.

File: pytorch_lightning/trainer/trainer.py [hunk 10]
```diff
@@ -318,6 +319,12 @@ def __init__(
         self.num_tpu_cores = num_tpu_cores
         assert num_tpu_cores in [1, 8, None], 'num_tpu_cores can only be 1 or 8'
 
+        if num_processes != 1 and distributed_backend != "ddp_cpu":
+            rank_zero_warn(
+                "num_processes is only used for distributed_backend=\"ddp_cpu\". Ignoring it."
+            )
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py:325]:
```suggestion
            rank_zero_warn("num_processes is only used for distributed_backend=\"ddp_cpu\". Ignoring it.")
```

File: pytorch_lightning/core/lightning.py [hunk 11]
```diff
@@ -944,11 +944,15 @@ def init_ddp_connection(self):
...
+            torch_backend = "nccl"
+        else:
+            torch_backend = "gloo"
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:954]:
```suggestion
        torch_backend = "nccl" if self.trainer.on_gpu else "gloo"
```
